clang/test/SemaSYCL/sycl-kernel-entry-point-attr-argument.cpp

```diff
@@ -209,8 +209,40 @@ struct B25_2;
   clang::sycl_kernel_entry_point(B25_2)]]
 void bad25();
 
-struct B26;
-// expected-error@+3 {{'sycl_kernel_entry_point' kernel name argument conflicts with a previous declaration}}
+struct B26 { void operator()() const {} };
+// expected-error@+3 {{'sycl_kernel_entry_point' kernel name 'B26' conflicts with a previous declaration}}
 // expected-note@+1  {{previous declaration is here}}
-[[clang::sycl_kernel_entry_point(B26)]] void bad26_1();
-[[clang::sycl_kernel_entry_point(B26)]] void bad24_2();
+[[clang::sycl_kernel_entry_point(B26)]] void bad26_1(B26 k) { k(); }
+[[clang::sycl_kernel_entry_point(B26)]] void bad24_2(B26 k) { k(); }
+
+
+struct kernel_name;
+
+template<typename KN, typename K>
+[[clang::sycl_kernel_entry_point(KN)]]
+void sycl_entry_point(K ker, int i) { ker(); }
+
+template<typename KN, typename K>
+void sycl_entry_point(K ker, long i) { ker(); }
+
+void proxy() {
+  int i = 0;
+  long l = 0;
+  // expected-error@-8 {{'sycl_kernel_entry_point' kernel name 'kernel_name' conflicts with a previous declaration}}
+  // expected-note-re@+4  {{in instantiation of function template specialization 'sycl_entry_point<kernel_name, (lambda at{{.*}}}}
+  // expected-note@-10  {{previous declaration is here}}
+  // expected-note-re@+1  {{in instantiation of function template specialization 'sycl_entry_point<kernel_name, (lambda at{{.*}}}}
+  sycl_entry_point<kernel_name>([]{}, i);
+  sycl_entry_point<kernel_name>([]{}, i);
+  // this doesn't trigger any error.
+  sycl_entry_point<kernel_name>([]{}, l);
```

Comment:
Would an error have been issued for this case prior to your changes? I can't test right now, but I suspect the answer is no. This call is an exact match for the second overload, so there isn't any reason for the first overload to have been selected anyway. I think the test cases above suffice to validate the SFINAE concerns since they verify that a "no matching function for call to ..." error is not issued.

Reply:
> Would an error have been issued for this case prior to your changes?

Yes

so if I have these 2 calls:

```cpp
  int i = 42;
  long l = 42;
  sycl_entry_point<kernel_name>([]{}, i);
  sycl_entry_point<kernel_name>([]{}, l);
```

the compiler produces on the current target branch

```shell-session
error: 'sycl_kernel_entry_point' kernel name argument conflicts with a previous declaration
    8 | void sycl_entry_point(K ker, int) {
      |      ^
note: in instantiation of function template specialization 'sycl_entry_point<kernel_name, (lambda at sycl-kernel-entry-point-attr-conflict.cpp:24:33)>' requested here
   24 |   sycl_entry_point<kernel_name>([]{}, l);
      |   ^
note: previous declaration is here
    8 | void sycl_entry_point(K ker, int) {
      |      ^
1 error generated.
```

But no error should be generated as the second is to a function without the attribute.

Reply:
Oh, wow. I guess that implies that `CheckSYCLEntryPointFunctionDecl()` is getting called during overload resolution? Even for a previously instantiated declaration?

Reply:
yes, the caller of `CheckSYCLEntryPointFunctionDecl ` is also used to check if template parameter substitution yield a valid decl and, in turn, allows to select viable functions for the overload resolution.

> Even for a previously instantiated declaration?

I don't think so as the lookup should retrieve the decl rather than create a new one, but we probably should test that :).